File: threadtracker-plugin/src/main/groovy/com/codoon/threadtracker/plugins/ThreadTrackerPlugin.java
package com.codoon.threadtracker.plugins;

import com.android.build.api.instrumentation.InstrumentationScope;
import com.android.build.api.variant.ApplicationAndroidComponentsExtension;
import com.android.build.gradle.AppPlugin;

import org.gradle.api.Plugin;
import org.gradle.api.Project;

import kotlin.Unit;

public class ThreadTrackerPlugin implements Plugin<Project> {
    @Override
    public void apply(Project project) {
        System.out.println("hello ThreadTrackerPlugin");

        project.getRootProject().getSubprojects().forEach(
                subProject -> {
                    PluginUtils.addProjectName(subProject.getName());
                    PluginUtils.projectPathList.add(subProject.getProjectDir().toString());
                }
        );
        project.getPlugins().withType(AppPlugin.class).configureEach(
                appPlugin -> {

                    //val androidComponents =
                    ApplicationAndroidComponentsExtension androidComponents =
                            project.getExtensions().getByType(ApplicationAndroidComponentsExtension.class);

                    // Registers a callback to be called, when a new variant is configured
                    androidComponents.onVariants(
                            androidComponents.selector().all(),
                            variant -> {
                                variant.getInstrumentation().transformClassesWith(
                                        ThreadClassVisitorFactory.class,
                                        InstrumentationScope.PROJECT,
                                        exampleParams -> {
                                            return Unit.INSTANCE;
                                        }
                                );
                            }
                    );
                }
        );
    }
}

File: threadtracker-plugin/src/main/kotlin/com/codoon/threadtracker/plugins/ThreadTrackerPlugin.kt
package com.codoon.threadtracker.plugins

import com.android.build.api.artifact.ScopedArtifact
import com.android.build.api.variant.AndroidComponentsExtension
import com.android.build.api.variant.ScopedArtifacts
import org.gradle.api.Plugin
import org.gradle.api.Project
import java.util.function.Consumer

open class ThreadTrackerPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        println("hello ThreadTrackerPlugin")

        project.rootProject.subprojects.forEach(
            Consumer { subProject: Project ->
                PluginUtils.addProjectName(subProject.name)
                PluginUtils.projectPathList.add(subProject.projectDir.toString())
                PluginUtils.dealAptFile()
            }
        )

        val ace = project.extensions.getByType(AndroidComponentsExtension::class.java)
        ace.onVariants { variant ->
            val taskProvider = project.tasks.register(
                "${variant.name}ModifyClasses", ModifyClassesTask::class.java
            )

            variant.artifacts
                .forScope(ScopedArtifacts.Scope.ALL)
                .use(taskProvider)
                .toTransform(
                    ScopedArtifact.CLASSES,
                    ModifyClassesTask::allJars,
                    ModifyClassesTask::allDirectories,
                    ModifyClassesTask::output
                )
        }
    }
}
